be/src/agent/task_worker_pool.cpp

```diff
@@ -616,8 +616,96 @@ void* TaskWorkerPool::_push_worker_thread_callback(void* arg_this) {
+                                                    std::vector<TTabletId>* error_tablet_ids) {
+    TaskWorkerPool* worker_pool_this = (TaskWorkerPool*)arg_this;
+    int64_t transaction_id = publish_version_req.transaction_id;
+    Status error_status = Status::OK();
+
+    // each partition
+    for (auto& par_ver_info : publish_version_req.partition_version_infos) {
+        int64_t partition_id = par_ver_info.partition_id;
+        // get all partition related tablets and check whether the tablet have the related version
+
+        map<TabletInfo, RowsetSharedPtr> tablet_related_rs;
+        StorageEngine::instance()->txn_manager()->get_txn_related_tablets(transaction_id, partition_id,
+                                                                          &tablet_related_rs);
+
+        TVersion version = par_ver_info.version;
+
+        // vector for tablet_info.
+        std::vector<TabletInfo> tablet_infos;
+        tablet_infos.reserve(tablet_related_rs.size());
+
+        // vector for tablet publishing version status, which collects the execution results of the correspoding tablet.
+        std::vector<Status> statuses(tablet_related_rs.size(), Status::OK());
+
+        size_t idx = 0;
+        // each tablet
+        for (auto& tablet_rs : tablet_related_rs) {
+            tablet_infos.push_back(tablet_rs.first);
+
+            // submit publishing tablet version task to the threadpool.
+            auto st = threadpool->submit_func([&worker_pool_this, &tablet_rs, &statuses, idx, &version, &transaction_id,
+                                               &partition_id]() {
+                const TabletInfo& tablet_info = tablet_rs.first;
+                const RowsetSharedPtr& rowset = tablet_rs.second;
+                auto& status = statuses[idx];
+                // if rowset is null, it means this be received write task, but failed during write
+                // and receive fe's publish version task
+                // this be must return as an error tablet
+                if (rowset == nullptr) {
+                    LOG(WARNING) << "Not found rowset of tablet: " << tablet_info.tablet_id << ", txn_id "
+                                 << transaction_id;
+                    status = Status::NotFound(fmt::format("Not found rowset of tablet: {}, txn_id: {}",
+                                                          tablet_info.tablet_id, transaction_id));
+                    return;
+                }
+                EnginePublishVersionTask engine_task(transaction_id, partition_id, version, tablet_info, rowset);
+
+                status = worker_pool_this->_env->storage_engine()->execute_task(&engine_task);
+                if (!status.ok())
+                    LOG(WARNING) << "failed to publish version for tablet, tablet_id " << tablet_info.tablet_id
+                                 << ", txn_id " << transaction_id << ", err: " << status;
+            });
+
+            if (!st.ok()) return st;
```

Comment:
consider large num of tablet_related_rs

Reply:
Retry is added here when the threadpool is full.